discovery/bootstrapper.go

```diff
@@ -124,102 +124,130 @@ func (c *ChannelGraphBootstrapper) SampleNodeAddrs(numAddrs uint32,
-				// If we haven't yet reached our limit, then
-				// we'll copy over the details of this node
-				// into the set of addresses to be returned.
-				tcpAddr, ok := nodeAddr.(*net.TCPAddr)
-				if !ok {
-					// If this isn't a valid TCP address,
-					// then we'll ignore it as currently
-					// we'll only attempt to connect out to
-					// TCP peers.
-					return nil
-				}
-
-				// At this point, we've found an eligible node,
-				// so we'll return early with our shibboleth
-				// error.
-				a = append(a, &lnwire.NetAddress{
-					IdentityKey: node.PubKey(),
-					Address:     tcpAddr,
-				})
+		// calculate a hash of this node's pubkey and our accumulator, which
+		// acts as its lottery number into our sampling
+		h := sha256.Sum256(
+			append(c.hashAccumulator[:], node.PubKey().SerializeCompressed()...),
+		)
+		// convert array to a slice
+		hash := h[:]
+
+		// determine whether this node should be included
+		if len(nodes) == 0 {
+			// always include the first node
+			nodes = append(nodes, node)
+			hashes = append(hashes, hash[:])
+
+		} else if uint32(len(nodes)) < numAddrs ||
+			bytes.Compare(hash, hashes[len(hashes)-1]) < 0 {
+			// Either this hash is smaller than the largest hash among the
+			// current lottery winners, or we are still doing an initial fill.
+			// So we should include this node as a winner. This is done by
+			// inserting at the end and bubbling up, to maintain sorted order.
+			nodes = append(nodes, node)
+			hashes = append(hashes, hash)
+
+			// bubble up by looping backwards until in sorted order
+			for i := len(nodes) - 2; i >= 0 && bytes.Compare(hash, hashes[i]) < 0; i-- {
+				// swap up
+				nodes[i+1] = nodes[i]
+				hashes[i+1] = hashes[i]
+				nodes[i] = node
+				hashes[i] = hash
 			}
 
-			c.tried[nID] = struct{}{}
-
-			return errFound
-		})
-		if err != nil && err != errFound {
-			return nil, err
+			// enforce only `numAddrs` winners of the lottery
+			if uint32(len(nodes)) > numAddrs {
+				nodes = nodes[:numAddrs]
+				hashes = hashes[:numAddrs]
+			}
 		}
 
-		return a, nil
+		return nil
+	})
+	if err != nil {
+		return nil, err
 	}
 
-	// We'll loop and sample new addresses from the graph source until
-	// we've reached our target number of outbound connections or we hit 50
-	// attempts, which ever comes first.
-	var (
-		addrs []*lnwire.NetAddress
-		tries uint32
-	)
-	for tries < 30 && uint32(len(addrs)) < numAddrs {
-		sampleAddrs, err := sampleAddrs()
-		if err != nil {
-			return nil, err
-		}
-
-		tries++
+	// we now have our lottery winners, place them all into the ignore list
+	for _, node := range nodes {
+		nID := autopilot.NewNodeID(node.PubKey())
+		c.tried[nID] = struct{}{}
+	}
 
-		// We'll now rotate our hash accumulator one value forwards.
-		c.hashAccumulator = sha256.Sum256(c.hashAccumulator[:])
+	// Select the first address from each node, plus any more to make up for
+	// a shortage in available nodes
+	shortage := numAddrs - uint32(len(nodes))
```

Comment:
Ideally, we should be sampling nodes rather than addresses since nodes can be reached at multiple addresses. We can just return all the addresses of the "winners" regardless of how many they have.